{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code for Noah's Hypermutation Project<br><br/>\n",
    "Run the code in this notebook to generate the files used for analysis and the figures themselves.  <br><br/>\n",
    "Code runs by referencing paths in the file: fileDirectory.txt\n",
    "The first section of this notebook describes how to create files for analysis.  The second part creates the figures.  The final files are already included so you can skip ahead if you would like"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Environment:<br>\n",
    "Python 2.7.15 |Anaconda custom (64-bit)| (default, Dec 14 2018, 13:10:39)<br>\n",
    "R version 3.5.1 (2018-07-02) -- \"Feather Spray\"<br>\n",
    "\n",
    "**Key packages:**<br>\n",
    "pandas 0.24.2<br>\n",
    "jupyter 1.0.0<br>\n",
    "jupyter-client 5.2.4<br>\n",
    "jupyter-console 5.2.0<br>\n",
    "jupyter-core 4.4.0<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "#TODO make \n",
    "sys.path.append('/Users/friedman/Desktop/hypermutationProjectFinal/scripts/utilityScripts')\n",
    "import configuration_util\n",
    "filePathDict = configuration_util.get_all_files_path_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr style=\"border:2px solid gray\"> </hr>\n",
    "\n",
    "## Files not created by me\n",
    "### Files\n",
    "* **CANCER_TYPE_INFO** information about TMB, cancer type and msi status, generated by Chai\n",
    "* **TCGA_CANCER_TYPE_INFO** information about TCGA cancer types, from cluster\n",
    "* **TCGA_MSI_SCORES** msi scores for TCGA from the cluster\n",
    "* **GENE_LENGTH_INFO** information about coding sequence lengths for genes, from cluster, originall ensemble?\n",
    "* **DEP_MAP_DATA** depmap scores for all genes.  Taken from https://ndownloader.figshare.com/files/22629068 -->achilles_gene_effect.csv.  CRISPR data\n",
    "* **PHASING_DATA** generated by Alex G, using his phasing pipeline on IMPACT data\n",
    "* **HOTSPOT_DATA** summary of all known hotspots from https://cancerhotspots.org\n",
    "* **MICROSATELLITE_INFORMATION** summary of all microsatellite sites in IMPACT panel.  Generated by Craig\n",
    "* **NUCLEOSOME_DYAD_POSITIONS** summary of nucleosome dyad positions, taken from Alex G, before that from Nuria's paper\n",
    "\n",
    "### Mafs\n",
    "* **IMPACT_BASE_MAF** impact filtered mutations from portal, annotated from Novemeber 2019\n",
    "* **IMPACT_BASE_MAF_WITH_SYNONYMOUS** impact unfiltered mutations from portal\n",
    "* **ALL_EXOME_MAF** maf combining all Exome recapture mutations and all TCGA mutations\n",
    "* ****\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr style=\"border:2px solid gray\"> </hr>\n",
    "\n",
    "# Generation of analysis files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files generated on the cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Signatures\n",
    "\n",
    "Uses the script mutationSigUtils in mode 'runMutSigs.'  This calls the mutation signatures script found here: https://github.com/mskcc/mutation-signatures.  This adds trinucleotide information (if needed) and then runs the mutation signature algorithm. \n",
    "\n",
    "To produce **IMPACT_SIGNATURE_DECOMPOSITIONS** run: <br>\n",
    "<code>python friedman/myUtils/mutationSigUtils.py --inputMaf path/to/impact/maf --outputDir path/to/output/dir --outputFilename exomeRecaptureSignatures.tsv --mode runMutSigs\n",
    "</code>\n",
    "\n",
    "To produce **TCGA_SIGNATURE_DECOMPOSITIONS** run: <br>\n",
    "<code>python friedman/myUtils/mutationSigUtils.py --inputMaf  --outputDir path/to/output/dir --outputFilename exomeRecaptureSignatures.tsv --mode runMutSigs\n",
    "</code>\n",
    "\n",
    "To produce **EXOME_RECAPTURE_SIGNATURE_DECOMPOSITIONS** run: <br>\n",
    "<code>python friedman/myUtils/mutationSigUtils.py --inputMaf /juno/work/ccs/gongy/megatron_Jan5th/Result/cohort_level/mut_somatic.maf --outputDir path/to/output/dir --outputFilename exomeRecaptureSignatures.tsv --mode runMutSigs\n",
    "</code>\n",
    "<br>\n",
    "To produce **CLONAL_VS_SUBCLONAL_MUTATIONAL_SIGNATURES** run: <br>\n",
    "<code>python friedman/myUtils/mutationSigUtils.py --inputMaf  --outputDir path/to/output/dir --outputFilename exomeRecaptureSignatures.tsv --mode runMutSigs\n",
    "</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All possible mutations in IMPACT panel and consequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#On the cluster run:\n",
    "#this command generates a vcf with all positions\n",
    "cmdAllVcfPos = '/opt/common/CentOS_7-dev/bin/bcftools mpileup --skip-indels --fasta-ref /ifs/depot/pi/resources/genomes/GRCh37/fasta/b37.fasta --regions-file /ifs/depot/pi/resources/roslin_resources/targets/IMPACT468/b37/IMPACT468_b37_targets.bed --output impact468_snps.vcf /ifs/res/pi/Proj_09221.6f74bdee-146b-11e9-a68d-645106ef9e4c/bam/s_C_P3U7DR_N001_d.Group3.rg.md.abra.printreads.bam'\n",
    "\n",
    "#this command adds in every possible alt allele\n",
    "cmdAddAltAlleles = 'python add_all_variants_to_vcf.py ~/impact468_snps.vcf ~/friedman/myAdjustedDataFiles/impact468_all_possible_snps.vcf'\n",
    "\n",
    "#use commands of the form below to cut vcf up into chromosome sized chunks\n",
    "'bcftools view /home/friedman/friedman/myAdjustedDataFiles/impact468_all_possible_snps_sorted.vcf.gz --regions 4 -o /juno/work/taylorlab/friedman/myAdjustedDataFiles/allPossibleMutationsInIMPACT/vcfs/chr4AllMuts.vcf'\n",
    "#convert it to a maf: (note writes a .sh file)\n",
    "\n",
    "#runs vcf2Maf on all possible Snps maf\n",
    "'python /juno/work/taylorlab/friedman/myUtils/run_vcf2Maf_on_all_possibleSnpsMaf.py'\n",
    "\n",
    "#Annotate mafs with the maf to maf script\n",
    "'python /juno/work/taylorlab/friedman/myUtils/run_maf2Maf_on_all_possibleSnpsMaf.py'\n",
    "\n",
    "#Results in one maf per chromosome of all possible mutations found in the following directory:\n",
    "#They are included in my project in the file: files/expectedMutationInfo/\n",
    "'/juno/work/taylorlab/friedman/myAdjustedDataFiles/allPossibleMutationsInIMPACT/annotatedMafs'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pentanucleotide annotated mafs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr style=\"border:2px solid gray\"> </hr>\n",
    "\n",
    "# Files generated locally"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classification of hypermutated tumors\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = ' '.join(['python', filePathDict['SCRIPT_DEFINE_HYPERMUTATION_THRESHOLDS'])\n",
    "os.popen(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculation of expected rate of hotspot, oncogenic and truncating mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate signature propensity to cause drivers\n",
    "cmdGet\n",
    "\n",
    "ALL_POSSIBLE_MUTATION_SUMMARY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Estimating the expected number of consequential mutations: Nucleotide context method**<br/>\n",
    "Enumerate all possible mutation mafs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identifying tumor subclones and copy number\n",
    "<br>\n",
    "On our compute cluster we ran the FACETs algorithm with the following parameters:?\n",
    "On the cluster we then summarize these results through a shared codebase: \n",
    "<code>python myUtils/create_cncf_or_rdata_file_list.py</code><br><br>\n",
    "then run <code>python prepare_maf_anno_commands.py</code><br><br>\n",
    "then run <code>/juno/work/taylorlab/friedman/myUtils/runMafAnno.sh</code> which calls myUtils/runAnnotateMaf.R<br><br> \n",
    "then combine results with \n",
    "<code>python /juno/work/taylorlab/friedman/myUtils/concat_maf_util.py /juno/work/taylorlab/friedman/myAdjustedDataFiles/mafAnnoRuns/annotatedMafs /juno/work/taylorlab/friedman/myAdjustedDataFiles/mafAnnoRuns/combined_cncf_annotated.maf</code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adjust clonality calls for hypermutated tumors with flat genomes:\n",
    "#this command will provide adjusted clonality calls for the cases with flat genome\n",
    "#this command takes ~1hr to run\n",
    "cmd = ' '.join(['python', filePathDict['SCRIPT_ADJUST_CLONALITY_CALLS'],\n",
    "                filePathDict['files/mafs/combined_cncf_annotated.maf'], filePathDict['IMPACT_MAF_WITH_ADJUSTED_CLONALITY_ANNOTATION'], 'IMPACT'])\n",
    "os.popen(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNDS ANALYSIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#todo make sure the input and output commands are correct here\n",
    "#make sure this script properly adjusts the file to run dnds\n",
    "cmd = ' '.join(['RScript', filePathDict['SCRIPT_RUN_DNDS_CV'], filePathDict['IMPACT_BASE_MAF_WITH_SYNONYMOUS']])\n",
    "\n",
    "#results are written to:\n",
    "''\n",
    "#TODO THE DNDS ANALYSIS script now has an annoying software bug\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation Atribution**<br><br/>\n",
    "Attributes mutations to signatures as done in Alex's paper.  We use different thresholds for attribution to account for the different characteristics of hypermutated cases.  Note this is currently not used but included here now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python /Users/friedman/Desktop/hypermutationProjectFinal/scripts/utilityScripts/attribute_mutations_to_signatures.py /Users/friedman/Desktop/hypermutationProjectFinal/files/infoFiles/Stratton_signatures30.txt . 1 10 /Users/friedman/Desktop/hypermutationProjectFinal/files/infoFiles/impactSignatureCalls_Nov20_2019.tsv attributedMutations.tsv agingIsAlwaysPresent doSmokingCorrection\n"
     ]
    }
   ],
   "source": [
    "cmd = ' '.join(['python', filePathDict['SCRIPT_ATTRIBUTE_MUTATIONS_TO_SIGNATURES'],\n",
    "                filePathDict['SIGNATURE_SPECTRUM'], '.', '1', '10',\n",
    "                filePathDict['IMPACT_SIGNATURE_DECOMPOSITIONS'], 'attributedMutations.tsv',\n",
    "                'agingIsAlwaysPresent', 'doSmokingCorrection'])\n",
    "\n",
    "print cmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phasing\n",
    "results are generated using methods described in Alex Gorelick's paper: https://www.nature.com/articles/s41586-020-2315-8\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Constructing Phylogenetic Trees**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate Figures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FIGURE 1**<br>\n",
    "Run all cells in <code>scripts/figure1/make_figure1.ipynb</code><br>\n",
    "Then run <code>scripts/figure1/plotFigure1.R</code>\n",
    "\n",
    "**FIGURE 1S**<br>\n",
    "Run all cells in <code>scripts/figure1/make_figure1_supplementary_figures.ipynb</code><br>\n",
    "Then run <code>scripts/figure1/plotSupplementaryFiguresFig1.R</code>\n",
    "\n",
    "**FIGURE 2**<br>\n",
    "Run all cells in <code>scripts/figure2/make_figure2.ipynb</code><br>\n",
    "Then run <code>scripts/figure2/plotFigure2.R</code>\n",
    "\n",
    "**FIGURE 2S**<br>\n",
    "Run all cells in <code>scripts/figure1/make_figure2_supplementary_figures.ipynb</code><br>\n",
    "Then run <code>scripts/figure1/plotSupplementaryFiguresFig2.R</code>\n",
    "\n",
    "**FIGURE 3**<br>\n",
    "Run all cells in <code>scripts/figure1/make_figure1.ipynb</code><br>\n",
    "Then run <code>scripts/figure1/plotFigure1.R</code>\n",
    "\n",
    "**FIGURE 3S**<br>\n",
    "Run all cells in <code>scripts/figure1/make_figure3_supplementary_figures.ipynb</code><br>\n",
    "Then run <code>scripts/figure1/plotSupplementaryFiguresFig3.R</code>\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
