{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code for Noah's Hypermutation Project<br><br/>\n",
    "Run the code in this notebook to generate the files used for analysis and the figures themselves.  <br><br/>\n",
    "Code runs by referencing paths in the file: fileDirectory.txt\n",
    "The first section of this notebook describes how to create files for analysis.  The second part creates the figures.  The final files are already included so you can skip ahead if you would like"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Environment:<br>\n",
    "Python 2.7.15 |Anaconda custom (64-bit)| (default, Dec 14 2018, 13:10:39)<br>\n",
    "R version 3.5.1 (2018-07-02) -- \"Feather Spray\"<br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "#TODO make \n",
    "sys.path.append('/Users/friedman/Desktop/hypermutationProjectFinal/scripts/utilityScripts')\n",
    "import configuration_util\n",
    "filePathDict = configuration_util.get_all_files_path_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr style=\"border:2px solid gray\"> </hr>\n",
    "\n",
    "## Files not created by me\n",
    "### Files\n",
    "* **CANCER_TYPE_INFO** information about TMB, cancer type and msi status, generated by Chai\n",
    "* **TCGA_CANCER_TYPE_INFO** information about TCGA cancer types, from cluster\n",
    "* **GENE_LENGTH_INFO** information about coding sequence lengths for genes, from cluster, originall ensemble?\n",
    "* **DEP_MAP_DATA** depmap scores for all genes.  Taken from https://ndownloader.figshare.com/files/22629068 -->achilles_gene_effect.csv.  CRISPR data\n",
    "* **PHASING_DATA** generated by Alex G, using his phasing pipeline on IMPACT data\n",
    "* **HOTSPOT_DATA** summary of all known hotspots from cancerhotspots.org\n",
    "* **MICROSATELLITE_INFORMATION** summary of all microsatellite sites in IMPACT panel.  Generated by Craig\n",
    "* **NUCLEOSOME_DYAD_POSITIONS** summary of nucleosome dyad positions, taken from Alex G, before that from Nuria's paper\n",
    "\n",
    "### Mafs\n",
    "* **IMPACT_BASE_MAF** impact filtered mutations from portal, annotated from Novemeber 2019\n",
    "* **IMPACT_BASE_MAF_WITH_SYNONYMOUS** impact unfiltered mutations from portal\n",
    "* **ALL_EXOME_MAF** maf combining all Exome recapture mutations and all TCGA mutations\n",
    "* ****\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr style=\"border:2px solid gray\"> </hr>\n",
    "\n",
    "# Generation of analysis files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files generated on the cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Signatures\n",
    "run <br>\n",
    "produces **IMPACT_SIGNATURE_DECOMPOSITIONS**\n",
    "run <br>\n",
    "produces **TCGA_SIGNATURE_DECOMPOSITIONS**\n",
    "run <br>\n",
    "produces **EXOME_RECAPTURE_SIGNATURE_DECOMPOSITIONS**\n",
    "run <br>\n",
    "produces **CLONAL_VS_SUBCLONAL_MUTATIONAL_SIGNATURES**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Calculation of expected rate of hotspot, oncogenic and truncating mutation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#On the cluster run:\n",
    "#command generates a vcf with all positions\n",
    "cmdAllVcfPos = '/opt/common/CentOS_7-dev/bin/bcftools mpileup --skip-indels --fasta-ref /ifs/depot/pi/resources/genomes/GRCh37/fasta/b37.fasta --regions-file /ifs/depot/pi/resources/roslin_resources/targets/IMPACT468/b37/IMPACT468_b37_targets.bed --output impact468_snps.vcf /ifs/res/pi/Proj_09221.6f74bdee-146b-11e9-a68d-645106ef9e4c/bam/s_C_P3U7DR_N001_d.Group3.rg.md.abra.printreads.bam'\n",
    "#add in every possible alt allele\n",
    "cmdAddAltAlleles = 'python add_all_variants_to_vcf.py ~/impact468_snps.vcf ~/friedman/myAdjustedDataFiles/impact468_all_possible_snps.vcf'\n",
    "#use commands of the form below to cut vcf up into chromosome sized chunks\n",
    "'bcftools view /home/friedman/friedman/myAdjustedDataFiles/impact468_all_possible_snps_sorted.vcf.gz --regions 4 -o /juno/work/taylorlab/friedman/myAdjustedDataFiles/allPossibleMutationsInIMPACT/vcfs/chr4AllMuts.vcf'\n",
    "#convert it to a maf: (note writes a .sh file)\n",
    "#run vcf2Maf\n",
    "python run_vcf2Maf_on_all_possibleSnpsMaf.py\n",
    "#annotateMafs\n",
    "python run_maf2Maf_on_all_possibleSnpsMaf.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pentanucleotide annotated mafs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Classification of hypermutated tumors**<br><br/>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = ' '.join(['python', filePathDict['SCRIPT_DEFINE_HYPERMUTATION_THRESHOLDS'])\n",
    "os.popen(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Calculation of expected rate of hotspot, oncogenic and truncating mutation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate signature propensity to cause drivers\n",
    "cmdGet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Estimating the expected number of consequential mutations: Nucleotide context method**<br/>\n",
    "Enumerate all possible mutation mafs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Identifying tumor subclones and copy number**<br><br/>\n",
    "On our compute cluster we ran the FACETs algorithm with the following parameters:?\n",
    "On the cluster we then summarize these results through a shared codebase: python myUtils/create_cncf_or_rdata_file_list.py then python prepare_maf_anno_commands.py then /juno/work/taylorlab/friedman/myUtils/runMafAnno.sh which calls myUtils/runAnnotateMaf.R then combine results with python concat_maf_util.py /juno/work/taylorlab/friedman/myAdjustedDataFiles/mafAnnoRuns/annotatedMafs /juno/work/taylorlab/friedman/myAdjustedDataFiles/mafAnnoRuns/combined_cncf_annotated.maf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr style=\"border:2px solid gray\"> </hr>\n",
    "\n",
    "## Files generated locally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adjust clonality calls for hypermutated tumors with flat genomes:\n",
    "#this command takes ~1hr to run\n",
    "cmd = ' '.join(['python', filePathDict['SCRIPT_ADJUST_CLONALITY_CALLS'],\n",
    "                filePathDict['files/mafs/combined_cncf_annotated.maf'], filePathDict['IMPACT_MAF_WITH_ADJUSTED_CLONALITY_ANNOTATION'], 'IMPACT'])\n",
    "os.popen(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**DNDS ANALYSIS**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#todo make sure the input and output commands are correct here\n",
    "#make sure this script properly adjusts the file to run dnds\n",
    "cmd = ' '.join(['RScript', filePathDict['SCRIPT_RUN_DNDS_CV'], filePathDict['IMPACT_BASE_MAF_WITH_SYNONYMOUS']])\n",
    "\n",
    "#results are written to:\n",
    "''\n",
    "#TODO THE DNDS ANALYSIS script now has an annoying software bug\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation Atribution**<br><br/>\n",
    "Attributes mutations to signatures as done in Alex's paper.  We use different thresholds for attribution to account for the different characteristics of hypermutated cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python /Users/friedman/Desktop/hypermutationProjectFinal/scripts/utilityScripts/attribute_mutations_to_signatures.py /Users/friedman/Desktop/hypermutationProjectFinal/files/infoFiles/Stratton_signatures30.txt . 1 10 /Users/friedman/Desktop/hypermutationProjectFinal/files/infoFiles/impactSignatureCalls_Nov20_2019.tsv attributedMutations.tsv agingIsAlwaysPresent doSmokingCorrection\n"
     ]
    }
   ],
   "source": [
    "cmd = ' '.join(['python', filePathDict['SCRIPT_ATTRIBUTE_MUTATIONS_TO_SIGNATURES'],\n",
    "                filePathDict['SIGNATURE_SPECTRUM'], '.', '1', '10',\n",
    "                filePathDict['IMPACT_SIGNATURE_DECOMPOSITIONS'], 'attributedMutations.tsv',\n",
    "                'agingIsAlwaysPresent', 'doSmokingCorrection'])\n",
    "#os.popen(cmd)\n",
    "\n",
    "#ALERT THIS IS BUGGY\n",
    "print cmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**PHASING**<br><br/>\n",
    "results are generated using methods described in: ALEX's paper"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Constructing Phylogenetic Trees**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate Figures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FIGURE 1**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#FIGURE 1A \n",
    "#FIGURE 1B\n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_1bc']\n",
    "#FIGURE 1C\n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_1bc']\n",
    "#FIGURE 1D\n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_1d']\n",
    "#FIGURE 1E\n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_1e']\n",
    "#FIGURE 1F\n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_1f']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FIGURE 2**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#FIGURE 2A \n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_2ac']\n",
    "#FIGURE 2B\n",
    "#FIGURE 2C\n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_2ac']\n",
    "#FIGURE 2D\n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_2d']\n",
    "#FIGURE 2E"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FIGURE 3**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#FIGURE 3A\n",
    "#FIGURE 3B\n",
    "#FIGURE 3C\n",
    "#FIGURE 3D\n",
    "#FIGURE 3E\n",
    "filePathDict['SCRIPT_GENERATE_PLOT_FIGURE_3e']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FIGURE 4**"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
