#written by noah friedman, help from craig to run
#a template for R scripts for plotting

library(ggplot2)
library(grid)
require(cowplot)
library(egg)
library(dplyr)
library(data.table); setDTthreads(6)
library(dndscv)
library(data.table)
library(ggrepel)
library(plyr)

fix_maf_cols <- function(maf){
  maf <- data.table(maf)
  mutations <- maf[Variant_Type %in% c("SNP", "INS", "DEL"),.
                   (sampleID = Tumor_Sample_Barcode,
                     chr = Chromosome,
                     pos = Start_Position,
                     ref = Reference_Allele,
                     mut = Tumor_Seq_Allele2, 
                     gene= Hugo_Symbol,
                     hgvs =HGVSp_Short,
                     status=hypermutantStatus,
                     cancerType=cancerType
                   )]
  #mutations$chr = paste("chr", as.vector(mutations$chr), sep='')
    #gsub("chr","",as.vector(mutations$chr))
  return(mutations)
}

run_dnds_on_normal_and_hypermutated_tumors <- function(inputMaf, cancerTypes){
  gene_list468 <- c('ABL1', 'ACVR1', 'AGO2', 'AKT1', 'AKT2', 'AKT3', 'ALK', 'ALOX12B', 'AMER1', 'ANKRD11', 'APC', 'AR', 'ARAF', 'ARID1A', 'ARID1B', 'ARID2', 'ARID5B', 'ASXL1', 'ASXL2', 'ATM', 'ATR', 'ATRX', 'AURKA', 'AURKB', 'AXIN1', 'AXIN2', 'AXL', 'B2M', 'BABAM1', 'BAP1', 'BARD1', 'BBC3', 'BCL10', 'BCL2', 'BCL2L1', 'BCL2L11', 'BCL6', 'BCOR', 'BIRC3', 'BLM', 'BMPR1A', 'BRAF', 'BRCA1', 'BRCA2', 'BRD4', 'BRIP1', 'BTK', 'CALR', 'CARD11', 'CARM1', 'CASP8', 'CBFB', 'CBL', 'CCND1', 'CCND2', 'CCND3', 'CCNE1', 'CD274', 'CD276', 'CD79A', 'CD79B', 'CDC42', 'CDC73', 'CDH1', 'CDK12', 'CDK4', 'CDK6', 'CDK8', 'CDKN1A', 'CDKN1B', 'CDKN2A', 'CDKN2B', 'CDKN2C', 'CEBPA', 'CENPA', 'CHEK1', 'CHEK2', 'CIC', 'CREBBP', 'CRKL', 'CRLF2', 'CSDE1', 'CSF1R', 'CSF3R', 'CTCF', 'CTLA4', 'CTNNB1', 'CUL3', 'CXCR4', 'CYLD', 'CYSLTR2', 'DAXX', 'DCUN1D1', 'DDR2', 'DICER1', 'DIS3', 'DNAJB1', 'DNMT1', 'DNMT3A', 'DNMT3B', 'DOT1L', 'DROSHA', 'DUSP4', 'E2F3', 'EED', 'EGFL7', 'EGFR', 'EIF1AX', 'EIF4A2', 'EIF4E', 'ELF3', 'EP300', 'EPAS1', 'EPCAM', 'EPHA3', 'EPHA5', 'EPHA7', 'EPHB1', 'ERBB2', 'ERBB3', 'ERBB4', 'ERCC2', 'ERCC3', 'ERCC4', 'ERCC5', 'ERF', 'ERG', 'ERRFI1', 'ESR1', 'ETV1', 'ETV6', 'EZH1', 'EZH2', 'FAM175A', 'FAM46C', 'FAM58A', 'FANCA', 'FANCC', 'FAT1', 'FBXW7', 'FGF19', 'FGF3', 'FGF4', 'FGFR1', 'FGFR2', 'FGFR3', 'FGFR4', 'FH', 'FLCN', 'FLT1', 'FLT3', 'FLT4', 'FOXA1', 'FOXL2', 'FOXO1', 'FOXP1', 'FUBP1', 'FYN', 'GATA1', 'GATA2', 'GATA3', 'GLI1', 'GNA11', 'GNAQ', 'GNAS', 'GPS2', 'GREM1', 'GRIN2A', 'GSK3B', 'GTF2I', 'H3F3A', 'H3F3B', 'H3F3C', 'HGF', 'HIST1H1C', 'HIST1H2BD', 'HIST1H3A', 'HIST1H3B', 'HIST1H3C', 'HIST1H3D', 'HIST1H3E', 'HIST1H3F', 'HIST1H3G', 'HIST1H3H', 'HIST1H3I', 'HIST1H3J', 'HIST2H3C', 'HIST2H3D', 'HIST3H3', 'HLA-A', 'HLA-B', 'HNF1A', 'HOXB13', 'HRAS', 'ICOSLG', 'ID3', 'IDH1', 'IDH2', 'IFNGR1', 'IGF1', 'IGF1R', 'IGF2', 'IKBKE', 'IKZF1', 'IL10', 'IL7R', 'INHA', 'INHBA', 'INPP4A', 'INPP4B', 'INPPL1', 'INSR', 'IRF4', 'IRS1', 'IRS2', 'JAK1', 'JAK2', 'JAK3', 'JUN', 'KDM5A', 'KDM5C', 'KDM6A', 'KDR', 'KEAP1', 'KIT', 'KLF4', 'KNSTRN', 'KRAS', 'LATS1', 'LATS2', 'LMO1', 'LYN', 'MALT1', 'MAP2K1', 'MAP2K2', 'MAP2K4', 'MAP3K1', 'MAP3K13', 'MAP3K14', 'MAPK1', 'MAPK3', 'MAPKAP1', 'MAX', 'MCL1', 'MDC1', 'MDM2', 'MDM4', 'MED12', 'MEF2B', 'MEN1', 'MET', 'MGA', 'MITF', 'MLH1','KMT2A', 'KMT2D', 'KMT2C', 'KMT2B', 'MPL', 'MRE11A', 'MSH2', 'MSH3', 'MSH6', 'MSI1', 'MSI2', 'MST1', 'MST1R', 'MTOR', 'MUTYH', 'MYC', 'MYCL1', 'MYCN', 'MYD88', 'MYOD1', 'NBN', 'NCOA3', 'NCOR1', 'NEGR1', 'NF1', 'NF2', 'NFE2L2', 'NFKBIA', 'NKX2-1', 'NKX3-1', 'NOTCH1', 'NOTCH2', 'NOTCH3', 'NOTCH4', 'NPM1', 'NRAS', 'NSD1', 'NTHL1', 'NTRK1', 'NTRK2', 'NTRK3', 'NUF2', 'NUP93', 'PAK1', 'PAK7', 'PALB2', 'PARK2', 'PARP1', 'PAX5', 'PBRM1', 'PDCD1', 'PDCD1LG2', 'PDGFRA', 'PDGFRB', 'PDPK1', 'PGR', 'PHOX2B', 'PIK3C2G', 'PIK3C3', 'PIK3CA', 'PIK3CB', 'PIK3CD', 'PIK3CG', 'PIK3R1', 'PIK3R2', 'PIK3R3', 'PIM1', 'PLCG2', 'PLK2', 'PMAIP1', 'PMS1', 'PMS2', 'PNRC1', 'POLD1', 'POLE', 'PPARG', 'PPM1D', 'PPP2R1A', 'PPP4R2', 'PPP6C', 'PRDM1', 'PRDM14', 'PREX2', 'PRKAR1A', 'PRKCI', 'PRKD1', 'PTCH1', 'PTEN', 'PTP4A1', 'PTPN11', 'PTPRD', 'PTPRS', 
                    'PTPRT', 'RAB35', 'RAC1', 'RAC2', 'RAD21', 'RAD50', 'RAD51', 'RAD51B', 'RAD51C', 'RAD51D', 'RAD52', 'RAD54L', 'RAF1', 'RARA', 'RASA1', 'RB1', 'RBM10', 'RECQL', 'RECQL4', 'REL', 'RET', 'RFWD2', 'RHEB', 'RHOA', 'RICTOR', 'RIT1', 'RNF43', 'ROS1', 'RPS6KA4', 'RPS6KB2', 'RPTOR', 'RRAGC', 'RRAS', 'RRAS2', 'RTEL1', 'RUNX1', 'RXRA', 'RYBP', 'SDHA', 'SDHAF2', 'SDHB', 'SDHC', 'SDHD', 'SESN1', 'SESN2', 'SESN3', 'SETD2', 'SETD8', 'SF3B1', 'SH2B3', 'SH2D1A', 'SHOC2', 'SHQ1', 'SLX4', 'SMAD2', 'SMAD3', 'SMAD4', 'SMARCA4', 'SMARCB1', 'SMARCD1', 'SMO', 'SMYD3', 'SOCS1', 'SOS1', 'SOX17', 'SOX2', 'SOX9', 'SPEN', 'SPOP', 'SPRED1', 'SRC', 'SRSF2', 'STAG2', 'STAT3', 'STAT5A', 'STAT5B', 'STK11', 'STK19', 'STK40', 'SUFU', 'SUZ12', 'SYK', 'TAP1', 'TAP2', 'TBX3', 'TCEB1', 'TCF3', 'TCF7L2', 'TEK', 'TERT', 'TET1', 'TET2', 'TGFBR1', 'TGFBR2', 'TMEM127', 'TMPRSS2', 'TNFAIP3', 'TNFRSF14', 'TOP1', 'TP53', 'TP53BP1', 'TP63', 'TRAF2', 'TRAF7', 'TSC1', 'TSC2', 'TSHR', 'U2AF1', 'UPF1', 'VEGFA', 'VHL', 'VTCN1', 'WHSC1', 'WHSC1L1', 'WT1', 'WWTR1', 'XIAP', 'XPO1', 'XRCC2', 'YAP1', 'YES1', 'ZFHX3', 'ZRSR2')
  impactGenesThatDontWorkForDnds <- c("MAP3K14", "MYCL1", "RYBP", "KMT5A", "PAK5", "FAM123B")
  gene_list468 <- setdiff(gene_list468, impactGenesThatDontWorkForDnds)
  
  dndsResults <- vector(mode="list", length=length(cancerTypes))
  names(dndsResults) <- cancerTypes
  
  #MAIN LOOP TO RUN DNDS on all cohorts
  for(cType in cancerTypes){
    print(paste('working on', cType))
    localList <- vector(mode="list", length=2)
    names(localList) <- c('Hypermutated', 'Normal')
   
    normalMaf <- inputMaf[(inputMaf$cancerType == cType) & (inputMaf$status == 'Normal'),]   #read.table(normalPath, sep = '\t', header=TRUE)
    hyperMaf <- maf[(inputMaf$cancerType == cType) & (inputMaf$status == 'Hypermutated'),]#read.table(hyperPath, sep = '\t', header=TRUE)
    
    normalMaf <- normalMaf[,c('sampleID', 'chr' ,'pos' ,'ref' ,'mut')]
    hyperMaf <- hyperMaf[,c('sampleID', 'chr' ,'pos' ,'ref' ,'mut')]
    
    print(paste('running normal, nmut:', dim(normalMaf)))
    dndsNormal <- dndscv(normalMaf, gene_list = gene_list468, max_muts_per_gene_per_sample = Inf, max_coding_muts_per_sample = Inf, use_indel_sites=F)
    print(paste('running hyper, nmut:', dim(hyperMaf)))
    dndsHyper <- dndscv(hyperMaf, gene_list = gene_list468, max_muts_per_gene_per_sample = Inf, max_coding_muts_per_sample = Inf, use_indel_sites=F)
    
    localList$Hypermutated <- dndsHyper
    localList$Normal <- dndsNormal
    dndsResults[[cType]] <- localList
  }
  
  return(dndsResults)
  
}

#TODO there is an error similar to:
#https://github.com/im3sanger/dndscv/issues/14
cancerTypes=
  c('Colorectal Cancer', 'Endometrial Cancer', 'Glioma',
    'Prostate Cancer', 'Esophagogastric Cancer', 'Bladder Cancer')
maf <- read.table('/Users/friedman/Desktop/hypermutationProjectFinal/files/mafs/impact_mutations_maf_synonymous_included.maf', sep = ',', header=TRUE)
maf <- fix_maf_cols(maf)
dndsResults <- run_dnds_on_normal_and_hypermutated_tumors(maf, cancerTypes)

dndsMerged <- vector(mode="list", length=length(cancerTypes))
for(cancerType in cancerTypes){
  sel_cvN <- dndsResults[[cancerType]]$Normal$sel_cv
  sel_cvH <- dndsResults[[cancerType]]$Hypermutated$sel_cv
  mergedCtype <- merge(sel_cvN, sel_cvH, suffixes = c(".Normal",".Hypermutated"), by='gene_name')
  mergedCtype$cancerType <- cancerType
  dndsMerged[[cancerType]] <- mergedCtype
}

combinedDf <- bind_rows(dndsMerged)

colnames(combinedDf)
plt <- make_dnds_plot(combinedDf, title='DNDS Hypermutated vs\nNon-hypermutated')
ggsave('~/Desktop/plot.pdf', plt,  width = 8, height = 8, units = c("in"))

write.table(combinedDf, file='/Users/friedman/Desktop/mnt/juno/work/taylorlab/friedman/myAdjustedDataFiles/dndsHypermutants.tsv', quote=FALSE, sep='\t')


####
########
##############
##################
########################
##################
##############
###########
#########
###

#TODO move this analysis to its own neighborhood
df <- read.table('/Users/friedman/Desktop/WORK/dataForLocalPlotting/dndscv_pancan_vs_local_genes.tsv', sep = '\t', header=TRUE)

endoNormalGenes <- unique(sel_cvEdnoN[sel_cvEdnoN$qglobal_cv < .01,]$gene_name)
coloNormalGenes <- unique(sel_cvColoN[sel_cvColoN$qglobal_cv < .01,]$gene_name)


#dfLim <- subset(df, !(gene %in% endoNormalGenes))
dfLim <- subset(df, !(gene %in% coloNormalGenes))
ggplot(dfLim, aes(x=nOncogenicMuts, y=1- log10(qVal), label=displayLabelOncogenic))+
  geom_point()+
  geom_text_repel()+
  geom_segment(aes(xend=0, yend=2, y=2, x=40))+
  geom_segment(aes(xend=0, yend=3, y=3, x=40))+
  xlim(0,50)+
  xlab('N oncogenic mutations across hypermutated cancer-type cohort')+
  ylab('1 - log10(pan-cancer dndscv q value)')+
  ggtitle('Oncogenic mutations in colorectal cancer\nin non-colorectal cancer related genes')



#PLOT THE DNDSCV information


df <- read.table('/Users/friedman/Desktop/WORK/dataForLocalPlotting/dndscvSummary.tsv', sep = '\t', header=TRUE)

make_dnds_plot(df, 'DNDS-CV Comparing Hypermutated and\n Non-Hypermutated Cancers')


